x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,177 @@
+	intermediateAccount := types.SuperfluidIntermediaryAccount{}
+	if wasSuperfluidDelegatedBeforeMigration {
+		// superfluid undelegate and break any underlying synthetic locks
+		// this is the same as SuperfluidUndelegate, but does not create a corresponding superunbonding synthetic lock
+		intermediateAccount, err = k.SuperfluidUndelegateToConcentratedPosition(ctx, sender.String(), lockId)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// Finish unlocking directly for locked or unlocking locks
+	err = k.lk.ForceUnlock(ctx, *lock)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// If shares to migrate is not specified, we migrate all shares.
+	if sharesToMigrate.IsZero() {
+		sharesToMigrate = gammSharesInLock
+	}
+
+	// Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
+	if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
+	}
+
+	// Exit the balancer pool position.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	// Defense in depth, ensuring we are returning exactly two coins.
+	if len(exitCoins) != 2 {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
+	}
+
+	// Create a full range (min to max tick) concentrated liquidity position.
+	if wasSuperfluidDelegatedBeforeMigration {
+		// If the lock was previously superfluid delegated, we create a new lock and keep it locked.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionLocked(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	} else {
+		// If the lock was unlocking, we create a new lock that is unlocking for the remaining time of the old lock.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was previously superfluid delegated, superfluid delegate the new concentrated lock to the same validator
+	if wasSuperfluidDelegatedBeforeMigration {
+		err := k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
```

Comment:
This can be merged with the "if" branch avove

Reply:
https://github.com/osmosis-labs/osmosis/pull/4885/commits/8df5fd0c9d0542dad2a9b426826f3bbfac62273f